File: apps/scala/src/main/scala/machinelearning/chapter08/Pipeline.scala
package machinelearning.chapter08

import machinelearning.chapter02.{Features, Row, Table, TrainTestSplit}

/** Survived.csv の特徴量の列と正解ラベルの列。 */
object SurvivedData:
  /** モデルに渡す特徴量の列 */
  val FeatureColumns: Vector[String] =
    Vector("Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked")

  /** 正解ラベルの列（1 が生存、0 が死亡） */
  val Target = "Survived"

  /** 行を、特徴量の列だけを並べた表にする。行がほかの列を持っていても使わない。 */
  def features(rows: Vector[Row]): Table = Table(FeatureColumns, rows)

  /** 行の Survived 列を、整数の正解ラベルにする。 */
  def target(rows: Vector[Row]): Vector[Int] = rows.map(_.text(Target).toInt)

/** 前処理を順に fit・transform してから、モデルを学習する。 */
case class Pipeline(transformers: Vector[Transformer], model: DecisionTreeClassifier):

  /** 訓練データで前処理とモデルを学習する。前処理は、前の前処理で変換したデータで fit する。 */
  def fit(x: Table, t: Vector[Int]): FittedPipeline =
    val (fitted, prepared) =
      transformers.foldLeft((Vector.empty[FittedTransformer], x)) {
        case ((fitted, prepared), transformer) =>
          val fittedTransformer = transformer.fit(prepared)
          (fitted :+ fittedTransformer, fittedTransformer.transform(prepared))
      }
    FittedPipeline(fitted, model.fit(FittedPipeline.toFeatures(prepared), t))

object Pipeline:
  /** Survived.csv 用のパイプライン。年齢・港の補完とダミー変数化の後に、決定木で分類する。 */
  def build(maxDepth: Option[Int], classWeight: ClassWeight): Pipeline =
    Pipeline(
      Vector(
        GroupMedianImputer("Age", Vector("Pclass", "Sex")),
        MostFrequentImputer("Embarked"),
        DummyEncoder(Vector("Sex", "Embarked"))
      ),
      DecisionTreeClassifier(maxDepth, classWeight)
    )

/** 学習済みの前処理とモデル。予測するときは前処理の transform だけを使う。 */
case class FittedPipeline(transformers: Vector[FittedTransformer], model: FittedDecisionTree):

  /** 学習済みの前処理を順に合成して、データを変換する。 */
  def transform(x: Table): Table =
    transformers.map(_.transform).foldLeft(identity[Table])(_ andThen _)(x)

  /** 前処理をして、モデルに渡す特徴量にする。 */
  def features(x: Table): Vector[Features] = FittedPipeline.toFeatures(transform(x))

  /** 前処理をしてから、1 行ごとのラベル（1 が生存、0 が死亡）を予測する。 */
  def predict(x: Table): Vector[Int] = model.predict(features(x))

object FittedPipeline:
  /** 前処理の済んだ表を特徴量にする。欠損値が残っていれば失敗する。 */
  private[chapter08] def toFeatures(x: Table): Vector[Features] =
    x.rows.map { row =>
      Features(
        x.columns,
        x.columns.map(column =>
          row.number(column).getOrElse(throw IllegalArgumentException(s"欠損値が残っています: $column"))
        )
      )
    }

/** 訓練データとテストデータの正解率と、テストデータの生存者のうち生存と予測できた人数。 */
case class Evaluation(
    trainAccuracy: Double,
    testAccuracy: Double,
    foundSurvivors: Int,
    survivors: Int
)

object Evaluation:
  private val Survived = 1

  /** 学習済みのパイプラインを、訓練データとテストデータで評価する。 */
  def evaluate(pipeline: FittedPipeline, split: TrainTestSplit[Row, Int]): Evaluation =
    val predictions = pipeline.predict(SurvivedData.features(split.xTest))
    val labels = split.tTest
    Evaluation(
      accuracy(pipeline.predict(SurvivedData.features(split.xTrain)), split.tTrain),
      accuracy(predictions, labels),
      predictions
        .zip(labels)
        .count((prediction, label) => prediction == Survived && label == Survived),
      labels.count(_ == Survived)
    )

  private def accuracy(predictions: Vector[Int], labels: Vector[Int]): Double =
    predictions.zip(labels).count((prediction, label) => prediction == label).toDouble / labels.size

File: apps/csharp/src/MachineLearning/Chapter08/Pipeline.cs
namespace MachineLearning.Chapter08;

using MachineLearning.Chapter02;

/// <summary>前処理を順に Fit・Transform してから、モデルを学習する。</summary>
public sealed record Pipeline(IReadOnlyList<ITransformer> Transformers, DecisionTreeClassifier Model)
{
    /// <summary>Survived.csv 用のパイプライン。年齢・港の補完とダミー変数化の後に、決定木で分類する。</summary>
    public static Pipeline Build(int maxDepth, ClassWeight classWeight) =>
        new(
            [
                new GroupMedianImputer("Age", ["Pclass", "Sex"]),
                new MostFrequentImputer("Embarked"),
                new DummyEncoder(["Sex", "Embarked"]),
            ],
            new DecisionTreeClassifier(maxDepth, classWeight));

    /// <summary>訓練データで前処理とモデルを学習する。前処理は、前の前処理で変換したデータで Fit する。</summary>
    public FittedPipeline Fit(Table x, IReadOnlyList<int> t)
    {
        ArgumentNullException.ThrowIfNull(x);
        var fitted = new List<IFittedTransformer>();
        var prepared = x;
        foreach (var transformer in this.Transformers)
        {
            var fittedTransformer = transformer.Fit(prepared);
            fitted.Add(fittedTransformer);
            prepared = fittedTransformer.Transform(prepared);
        }

        return new FittedPipeline(fitted, this.Model.Fit(FittedPipeline.FeaturesOf(prepared), t));
    }
}

/// <summary>学習済みの前処理とモデル。予測するときは Fit で覚えた値だけを使う。</summary>
public sealed record FittedPipeline(IReadOnlyList<IFittedTransformer> Transformers, FittedDecisionTree Model)
{
    /// <summary>学習済みの前処理を順に合成して、データを変換する。</summary>
    public Table Transform(Table x) => FittedTransformers.Compose(this.Transformers).Transform(x);

    /// <summary>前処理をして、モデルに渡す特徴量にする。</summary>
    public IReadOnlyList<Features> ToFeatures(Table x) => FeaturesOf(this.Transform(x));

    /// <summary>前処理をしてから、1 行ごとのラベル（1 が生存、0 が死亡）を予測する。</summary>
    public IReadOnlyList<int> Predict(Table x) => this.Model.Predict(this.ToFeatures(x));

    /// <summary>乗客 1 人分の行から、生存（1）か死亡（0）かを予測する。</summary>
    public int PredictOne(Row row) => this.Predict(SurvivedData.ToTable([row]))[0];

    /// <summary>前処理の済んだ表を特徴量にする。欠損値が残っていれば失敗する。</summary>
    internal static IReadOnlyList<Features> FeaturesOf(Table x) =>
        [.. x.Rows.Select(row => new Features(
            x.Columns,
            [.. x.Columns.Select(column => row.Number(column)
                ?? throw new ArgumentException($"欠損値が残っています: {column}", nameof(x)))]))];
}
